### PR TITLE
DTH: Validate dependency when search paths are updated

### PR DESCRIPTION
**Pull request by troydai**
/cc @davidfowl @abpiskunov @muratg @anurse 

Addressing #2671

### CODE REVIEW

File: src/Microsoft.Dnx.DesignTimeHost/ProjectStateResolver.cs [hunk 1]
```diff
@@ -72,9 +78,18 @@ public ProjectState Resolve(string appPath, string configuration, bool triggerBu
 
             var sourcesProjectWidesources = project.Files.SourceFiles.ToList();
 
+            ResolveSearchPaths(state);
+
+            var searchPathUpdated = currentSearchPaths != null &&
```
Comment (troydai) [inline src/Microsoft.Dnx.DesignTimeHost/ProjectStateResolver.cs]: Explaination: `currentSearchPath` is `null` when the project hasn't been initialized, which means it is processing the `Initialization` request now. There is no sense to validate the dependencies validation under changed search paths in this situation.

File: src/Microsoft.Dnx.DesignTimeHost/ProjectStateResolver.cs [hunk 2]
```diff
@@ -257,6 +256,73 @@ private DependencyInfo ResolveProjectDependencies(Project project, string config
...
+
+                var newErros = new List<DiagnosticMessageView>(description.Errors);
+                newErros.Add(new DiagnosticMessageView(newDiagnostic));
+                description.Errors = newErros;
+
+                diagnostics.Add(newDiagnostic);
+            }
+        }
+
+        private static void ResolveSearchPaths(ProjectState state)
```
Comment (troydai) [inline src/Microsoft.Dnx.DesignTimeHost/ProjectStateResolver.cs]: https://github.com/aspnet/dnx/issues/2793

File: src/Microsoft.Dnx.DesignTimeHost/ProjectStateResolver.cs [hunk 3]
```diff
@@ -257,6 +256,73 @@ private DependencyInfo ResolveProjectDependencies(Project project, string config
...
+                var newDiagnostic = new DiagnosticMessage(
+                    DiagnosticMonikers.NU1001,
+                    $"The dependency {description.Name} could not be resolved since the search paths are updated.",
+                    library.RequestedRange.FileName,
+                    DiagnosticMessageSeverity.Error,
+                    library.RequestedRange.Line,
+                    library.RequestedRange.Column,
+                    library);
+
+                var newErros = new List<DiagnosticMessageView>(description.Errors);
```
Comment (muratg) [inline src/Microsoft.Dnx.DesignTimeHost/ProjectStateResolver.cs]: Tpyo: `Errors`

**Comment by troydai** [comment]
People, this is a beta-8. I want more people helping me take a look and sign off this soon.

@abpiskunov @anurse @BrennanConroy @davidfowl @JunTaoLuo @moozzyk  @muratg

**Comment by anurse** [comment]
*cough* https://github.com/aspnet/dnx/pull/2795 *cough* :P

**Comment by troydai** [comment]
Wew, Wew, Wew, on my way to #2795

File: src/Microsoft.Dnx.DesignTimeHost/ProjectStateResolver.cs (hunk 1 above)
Comment (anurse) [inline src/Microsoft.Dnx.DesignTimeHost/ProjectStateResolver.cs]: Sounds like that would make a good comment ;)

**Comment by anurse** [comment]
Looks good to me. I'm not super confident I follow all the DTH logic involved though, so maybe @davidfowl should take a scan before you merge :).

File: src/Microsoft.Dnx.DesignTimeHost/ProjectStateResolver.cs [hunk 4]
```diff
@@ -121,48 +136,26 @@ public ProjectState Resolve(string appPath, string configuration, bool triggerBu
...
+                                                          string configuration,
+                                                          FrameworkName frameworkName,
+                                                          int protocolVersion,
+                                                          List<string> updatedSearchPath)
         {
             var cacheKey = Tuple.Create("DependencyInfo", project.Name, configuration, frameworkName);
 
             return _compilationEngine.CompilationCache.Cache.Get<DependencyInfo>(cacheKey, ctx =>
             {
+                var projectCandiates = updatedSearchPath?.Where(path => Directory.Exists(path))
```
Comment (davidfowl) [inline src/Microsoft.Dnx.DesignTimeHost/ProjectStateResolver.cs]: Make this a hashset.

File: src/Microsoft.Dnx.DesignTimeHost/ProjectStateResolver.cs [hunk 5]
```diff
@@ -257,6 +256,73 @@ private DependencyInfo ResolveProjectDependencies(Project project, string config
             });
         }
 
+        private void ValidateDependency(LibraryDescription library, DependencyDescription description, List<DiagnosticMessage> diagnostics, Dictionary<string, string> projectCandidates)
```
Comment (davidfowl) [inline src/Microsoft.Dnx.DesignTimeHost/ProjectStateResolver.cs]: Why not just modify the LibraryDescription?

File: src/Microsoft.Dnx.DesignTimeHost/ProjectStateResolver.cs [hunk 6]
```diff
@@ -257,6 +256,73 @@ private DependencyInfo ResolveProjectDependencies(Project project, string config
...
+            var foundCandidate = projectCandidates.TryGetValue(description.Name, out newPath);
+
+            if ((description.Type == LibraryTypes.Project && !foundCandidate) ||
+                (description.Type == LibraryTypes.Package && foundCandidate))
+            {
+                description.Resolved = false;
+                description.Type = LibraryTypes.Unresolved;
+
+                var newDiagnostic = new DiagnosticMessage(
+                    DiagnosticMonikers.NU1001,
```
Comment (davidfowl) [inline src/Microsoft.Dnx.DesignTimeHost/ProjectStateResolver.cs]: Use a different error code.

Comment (troydai) [inline src/Microsoft.Dnx.DesignTimeHost/ProjectStateResolver.cs]: Add new error code for library type change.

File: src/Microsoft.Dnx.DesignTimeHost/ProjectStateResolver.cs [hunk 7]
```diff
@@ -184,6 +177,12 @@ private DependencyInfo ResolveProjectDependencies(Project project, string config
                     var diagnostics = diagnosticSources[library];
                     var description = CreateDependencyDescription(library, diagnostics, protocolVersion);
 
+                    if (projectCandiates != null)
+                    {
+                        // Validate
+                        ValidateDependency(library, description, info.Diagnostics, projectCandiates);
```
Comment (troydai) [inline src/Microsoft.Dnx.DesignTimeHost/ProjectStateResolver.cs]: Exploring the possibility of moving it ahead of `CreateDependencyDescription`

**Comment by BrennanConroy** [comment]
Doesn't build on Travis

File: src/Microsoft.Dnx.DesignTimeHost/ProjectStateResolver.cs [hunk 8]
```diff
@@ -121,48 +136,23 @@ public ProjectState Resolve(string appPath, string configuration, bool triggerBu
...
+                                                          string configuration,
+                                                          FrameworkName frameworkName,
+                                                          int protocolVersion,
+                                                          List<string> updatedSearchPath)
         {
             var cacheKey = Tuple.Create("DependencyInfo", project.Name, configuration, frameworkName);
 
             return _compilationEngine.CompilationCache.Cache.Get<DependencyInfo>(cacheKey, ctx =>
             {
+
```
Comment (davidfowl) [inline src/Microsoft.Dnx.DesignTimeHost/ProjectStateResolver.cs]: remove spaces

File: src/Microsoft.Dnx.DesignTimeHost/ProjectStateResolver.cs [hunk 9]
```diff
@@ -177,11 +167,17 @@ private DependencyInfo ResolveProjectDependencies(Project project, string config
                     Diagnostics = libraryManager.GetAllDiagnostics().ToList()
                 };
 
-                var diagnosticSources = info.Diagnostics.ToLookup(diagnostic => diagnostic.Source);
```
Comment (davidfowl) [inline src/Microsoft.Dnx.DesignTimeHost/ProjectStateResolver.cs]: keep this and do a concat with the newly added diagnostic

**Comment by troydai** [comment]
Rebased and squashed. Waiting for Travis result

**Comment by davidfowl** [comment]
:shipit:

**Comment by troydai** [comment]
Travis passed. I'll merge shortly.

**Pull request merged**